### PR TITLE
eval_ppl for profile memory model

### PR DESCRIPTION
**Pull request by emilydinan**
Code for evaluating profilememory model. Using this code, the perplexity evaluations are as follows for the ConvAI2 and personachat test sets:

- personachat test: {'total': 7512, 'loss': 3.7, 'ppl': 40.42}
- convai2 test: {'total': 6634, 'loss': 3.81, 'ppl': 45.19}

### CODE REVIEW

File: projects/convai2/baselines/profilememory/eval_ppl.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+from parlai.core.thread_utils import SharedTable
+from parlai.core.utils import Timer, round_sigfigs, no_lock
+from parlai.core.worlds import World, create_task
+from projects.personachat.persona_seq2seq import PersonachatSeqseqAgentSplit
+import torch.nn.functional as F
+from torch.autograd import Variable
+from torch.autograd import Variable
+import torch
+import math
+import random
```
Comment (alexholdenmiller) [inline projects/convai2/baselines/profilememory/eval_ppl.py]: probably don't need all this, at least don't need `Variable` twice :)

File: projects/convai2/baselines/profilememory/eval_ppl.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+                dist[self.dict[i+1]] = probs[0][i]
+        return dist
+
+
+if __name__ == '__main__':
+    parser = setup_args()
+    parser.add_argument('-n', '--num-examples', default=100000000)
+    parser.add_argument('-d', '--display-examples', type='bool', default=False)
+    parser.add_argument('-ltim', '--log-every-n-secs', type=float, default=2)
+    PersonachatSeqseqAgentSplit.add_cmdline_args(parser)
```
Comment (alexholdenmiller) [inline projects/convai2/baselines/profilememory/eval_ppl.py]: don't need lines 145-148

File: projects/convai2/baselines/profilememory/eval_ppl.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            new_output = F.dropout(output.squeeze(0), p=self.dropout, training=False)
+            if self.sharelt:
+                e = self.h2e(new_output)
+                scores = self.e2o(e)[:,1:]
+            else:
+                scores = self.h2o(new_output)
+            probs = F.softmax(scores)
+            _max_score, preds = scores.max(1)
+
+            if zs is not None:
```
Comment (alexholdenmiller) [inline projects/convai2/baselines/profilememory/eval_ppl.py]: nothing in this if block is used, right? since probs is already used above. and don't need the _max_score, preds = scores.max(1) either

File: projects/convai2/baselines/profilememory/eval_ppl.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        x = Variable(self.START_TENSOR)
+        xe = self.lt(x).unsqueeze(1)
+        xes = xe.expand(xe.size(0), 1, xe.size(2))
+
+        attn_mask = Variable(xs_persona.data.sum(2).ne(0), requires_grad=False)
+
+        output_lines = [[] for _ in range(1)]
+
+        # we only ever want to generate the next token
+        self.longest_label = 1
```
Comment (alexholdenmiller) [inline projects/convai2/baselines/profilememory/eval_ppl.py]: can delete this

File: projects/convai2/baselines/profilememory/eval_ppl.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        if len(partial_out) > 0:
+            obs[0]['eval_labels'] = (' '.join(partial_out),)
+        else:
+            obs[0]['eval_labels'] = ['']
+
+        xs, xs_persona, ys, labels, valid_inds, cands, valid_cands, zs, eval_labels = self.batchify(obs)
+
+        self.encoder.train(mode=False)
+        self.encoder_persona.train(mode=False)
+        self.decoder.train(mode=False)
```
Comment (alexholdenmiller) [inline projects/convai2/baselines/profilememory/eval_ppl.py]: more clear to do `X.eval()`

File: projects/convai2/baselines/profilememory/eval_ppl.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        score for that word. Unset keys assume a probability of zero.
+        e.g.
+        {'text': 'Run test program.'}, ['hello'] => {'world': 1.0}
+        """
+        if not hasattr(self, 'last_text'):
+            self.last_text = None
+
+        if observation['text'] != self.last_text:
+            self.last_text = observation.get('text')
+            self.observe(observation)
```
Comment (alexholdenmiller) [inline projects/convai2/baselines/profilememory/eval_ppl.py]: can remove all the `self.last_text` stuff, not being used below

File: projects/convai2/baselines/profilememory/eval_ppl.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                xes = F.dropout(xes, p=self.dropout, training=False)
+                for b in range(1):
+                    # convert the output scores to tokens
+                    token = self.v2t([(preds+1).data[b]])
+                    output_lines[b].append(token)
+
+        if random.random() < 0.01 and zs is not None:
+            # sometimes output a prediction for debugging
+            print('label:', self.dict.vec2txt(zs.data[0]),
+                  '\nprediction:', output_lines[0][-1])
```
Comment (alexholdenmiller) [inline projects/convai2/baselines/profilememory/eval_ppl.py]: probably could cut the output_lines and self.v2t stuff too for simplicity

File: projects/convai2/baselines/profilememory/eval_ppl.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+        # extract probability distribution
+        probs = probs.tolist()
```
Comment (alexholdenmiller) [inline projects/convai2/baselines/profilememory/eval_ppl.py]: since you called tolist, you can remove the hasattr(probs, 'item') stuff below

File: projects/convai2/baselines/profilememory/eval_ppl.py (hunk 4 above)
Comment (emilydinan) [reply projects/convai2/baselines/profilememory/eval_ppl.py]: i think i need this?

File: projects/convai2/baselines/profilememory/eval_ppl.py (hunk 6 above)
Comment (emilydinan) [reply projects/convai2/baselines/profilememory/eval_ppl.py]: it is used to check whether or not we already observed the text

File: projects/convai2/baselines/profilememory/eval_ppl.py (hunk 3 above)
Comment (emilydinan) [reply projects/convai2/baselines/profilememory/eval_ppl.py]: yes it is used: we get a new `xes` by feeding the true token. i was using the `preds` to generate the output lines for debugging, etc. i can delete the `output_lines` code if you want but i thought it was helpful to see what the model was predicting.

File: projects/convai2/baselines/profilememory/eval_ppl.py (hunk 6 above)
Comment (alexholdenmiller) [reply projects/convai2/baselines/profilememory/eval_ppl.py]: ahhh yes I see I see

File: projects/convai2/baselines/profilememory/eval_ppl.py (hunk 3 above)
Comment (alexholdenmiller) [reply projects/convai2/baselines/profilememory/eval_ppl.py]: right yeah it's only for the output lines, sure we can keep if you like them
